types/arangodb/index.d.ts

```diff
@@ -1163,6 +1165,10 @@ declare namespace Foxx {
     interface Request {
         arangoUser: string | null;
         arangoVersion: number;
+        auth: null | {
+            bearer?: string;
+            basic?: { username?: string; password?: string };
+        };
```

Comment:
These are strictly mutually exclusive but `{bearer?: string} | {basic?: ...}` has unintuitive DX. I tried using `{bearer?: string, basic: never} | {bearer: never, basic?: ...}` but that didn't seem to guarantee exclusivity either.

Reply:
A union of two object types represents a type with all the properties of the first object type or all the properties of the second object type. However, this means that an object with all the properties of both objects satisfies this also.

In the definition with `never`, the error you're seeing is due to making the `never` typed property required instead of optional. If you change it a little, then it will work: [playground link](https://www.typescriptlang.org/play/index.html#src=type%20t%20%3D%20null%20%7C%20%7B%20bearer%3A%20string%2C%20basic%3F%3A%20never%20%7D%20%7C%20%7B%20bearer%3F%3A%20never%2C%20basic%3A%20%7B%20username%3F%3A%20string%2C%20password%3F%3A%20string%20%7D%20%7D%3B%0D%0A%0D%0Alet%20a%3A%20t%20%3D%20null%3B%20%2F%2F%20OK%0D%0Alet%20b%3A%20t%20%3D%20%7B%20bearer%3A%20%22b%22%20%7D%3B%20%2F%2FOK%0D%0Alet%20c%3A%20t%20%3D%20%7B%20basic%3A%20%7B%7D%20%7D%3B%20%2F%2F%20OK%0D%0Alet%20d%3A%20t%20%3D%20%7B%20bearer%3A%20%22d%22%2C%20basic%3A%20%7B%7D%20%7D%3B%20%2F%2F%20ERROR).

Reply:
@PranavSenthilnathan This is a return value though. The logic I was trying to express is "if bearer is present, basic is undefined; if basic is present, bearer is undefined" and "if basic.username is undefined, so is basic.password".

Typing this as `{ bearer: string } | { basic: { username?: string, password?: string } }` means I can't even check for existence of either property, which is awkward for users.

Typing this with `never` means the following still works:

```js
if (auth.bearer) {
  doSomething(auth.basic); // this property shouldn't be valid here
}
```

So it's effectively indistinguishable from the much simpler typing I've ended up with.

Reply:
If `doSomething` is typed to not accept `undefined` then you would get an error instead. I see what you're saying though - even though the property will be strongly typed as `unknown`, the property access will not be an error as it would be for accessing nonexistent properties on normal objects